lib/rust/parser/src/syntax/token.rs

```diff
@@ -0,0 +1,316 @@
+//! Token implementation. Tokens are output from the lexer. Read the docs of the main module to
+//! learn more about the parsing process steps.
+
+use crate::prelude::*;
+
+use crate::source::span::Span;
+use crate::source::span::SpanRef;
+use crate::source::Code;
+use crate::source::Offset;
+
+use enso_shapely_macros::tagged_enum;
+
+
+
+// =============
+// === Token ===
+// =============
+
+/// A lexical token is a string with an assigned and thus identified meaning. Token remembers its
+/// source code and can be printed back. It also contains information about the offset to the
+/// previous token if any.
+///
+/// There are many variants of tokens, however, some places in the code need to distinguish them,
+/// while some need to store several variants in the same collection. Because of that this module
+/// defines a lot of utilities for managing tokens and their variants.
+///
+/// # Variant structs
+/// The most primitive structure is the [`Variant`] one, which is an enum with many variants:
+///
+/// ```text
+/// pub enum Variant {
+///     Newline (variant::Newline),
+///     Symbol (variant::Symbol),
+///     Wildcard (variant::Wildcard),
+///     Ident (variant::Ident),
+///     // ... many more
+/// }
+/// ```
+///
+/// Every variant is a separate struct defined in the [`variant`] module. Every variant is
+/// associated with a constructor of the same name (tuple-struct like). For example, the
+/// [`variant::Ident`] is defined as:
+///
+/// ```text
+/// pub mod variant {
+///     pub struct Ident {
+///         pub is_free: bool,
+///         pub lift_level: usize
+///     }
+///     pub fn Ident(is_free: bool, lift_level: usize) -> Ident { ... }
+///     // ... many more variants
+/// }
+/// ```
+///
+/// There is a conversion defined between every of the variants and the [`Variant`] struct.
+/// Moreover, the [`Variant`] struct defines a constructor function for each of its variants. For
+/// example, the identifier variant constructor is defined as:
+///
+/// ```text
+/// impl Variant {
+///     pub fn ident(is_free: bool, lift_level: usize) -> Self {
+///         Self::Ident(variant::Ident(is_free, lift_level))
+///     }
+/// }
+/// ```
+///
+/// # Generic token type
+/// The [`Token`] structure has a default parametrization of [`Token<'s, Variant>`] which basically
+/// is a token containing any of the defined variants.
+///
+/// # Token variants
+/// The [`Token`] structure can be parametrized with a variant struct to form a token variant. This
+/// module defines type aliases for every such a combination. For example, the [`Ident`] token
+/// variant is defined as:
+///
+/// ```text
+/// pub type Ident<'s> = Token<'s, variant::Ident>;
+/// ```
+///
+/// There is a conversion defined between any [`Token<'s, T>`] and [`Token<'s>`] for [`T`] being
```

Comment:
It'd be nice to have also TryFrom for downcasting.

Reply:
Skipped it for now. The lifetime requirements are a mess and require the proc macro to be much more complicated. I'm not even sure if we can auto-generate it :(